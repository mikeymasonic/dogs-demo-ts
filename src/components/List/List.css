.List input[type='radio'] {
  display: none;
}

.List label {
  display: flex;
  color: powderblue;
  font-weight: bolder;
  font-size: 1rem;
  width: 15rem;
  height: 3rem;
  max-width: 15rem;
  margin: 0.5rem;
  align-items: center;
  justify-content: center;
  border: dotted 0.1vw rgb(197, 197, 197);
  background-color: purple;
  cursor: pointer;
}

.List label:hover {
  background: #004c97;
  color: greenyellow;
  font-weight: bolder;
  width: 15rem;
  height: 3rem;
  margin: 0.5rem;
}

.List input[type='radio']:checked + label {
  background: #004c97;
  color: greenyellow;
  width: 15rem;
}
