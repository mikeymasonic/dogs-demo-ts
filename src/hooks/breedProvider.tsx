import React, { createContext, useState, useContext, useEffect } from 'react';
import { fetchList, fetchDetail } from '../services/dogsAPI';

export type BreedContext = {
  breeds: Array<string>;
  filteredBreeds: Array<string>;
  searchText: string;
  breedDetail: Array<string>;
  loading: boolean;
  handleChange: (a: { target: any }) => void;
  handleDetail: (breed: string) => void;
};

const BreedContext = createContext<BreedContext | null>(null);

export const BreedProvider = ({ children }: { children: React.ReactNode }) => {
  const [breeds, setBreeds] = useState<string[]>([]);
  const [filteredBreeds, setFilteredBreeds] = useState<string[]>([]);
  const [searchText, setSearchText] = useState('');
  const [breedDetail, setBreedDetail] = useState<string[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setLoading(true);
    fetchList()
      .then((apiBreeds) => setBreeds(apiBreeds))
      .finally(() => setLoading(false));
  }, []);

  useEffect(() => {
    setFilteredBreeds(breeds.slice(0, 12));
  }, [breeds]);

  useEffect(() => {
    const searchResults = breeds.filter((breed) => breed.includes(searchText));
    setLoading(true);
    setFilteredBreeds(searchResults.slice(0, 12));
    setLoading(false);
  }, [breeds, searchText]);

  const handleChange = ({ target }: { target: any }) => {
    setSearchText(target.value);
  };

  const handleDetail = (breed: string) => {
    setLoading(true);
    fetchDetail(breed)
      .then((apiBreeds) => setBreedDetail(apiBreeds))
      .finally(() => setLoading(false));
  };

  return (
    <BreedContext.Provider
      value={{
        breeds,
        filteredBreeds,
        searchText,
        breedDetail,
        handleChange,
        handleDetail,
        loading,
      }}
    >
      {children}
    </BreedContext.Provider>
  );
};

export const useBreeds = () => {
  return useContext(BreedContext)?.breeds;
};

export const useFilteredBreeds = () => {
  return useContext(BreedContext)?.filteredBreeds;
};

export const useSearchText = () => {
  return useContext(BreedContext)?.searchText;
};

export const useBreedDetail = () => {
  return useContext(BreedContext)?.breedDetail;
};

export const useHandleChange = () => {
  return useContext(BreedContext)?.handleChange;
};

export const useHandleDetail = () => {
  return useContext(BreedContext)?.handleDetail;
};

export const useLoading = () => {
  return useContext(BreedContext)?.loading;
};

export default BreedContext;
