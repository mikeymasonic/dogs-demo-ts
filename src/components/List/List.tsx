import React from 'react';
import { useHandleDetail } from '../../hooks/breedProvider';

const List = ({ breed }: { breed: string }) => {
  const handleDetail = useHandleDetail();

  return (
    <section>
      <input
        type="radio"
        value={breed}
        name="dogs"
        id={breed}
        onClick={() => handleDetail!(breed)}
      />
      <label htmlFor={breed}>{breed}</label>
    </section>
  );
};

export default List;
