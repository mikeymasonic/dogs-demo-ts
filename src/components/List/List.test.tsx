import React from 'react';
import { shallow } from 'enzyme';
import * as BreedContext from '../../hooks/breedProvider';
import List from './List';

describe('List component', () => {
  test('it should mock the context and render List', () => {
    jest
      .spyOn(BreedContext, 'useHandleDetail')
      .mockImplementation(() => () => {});

    const wrapper = shallow(<List breed="husky" />);
    expect(wrapper).toMatchSnapshot();
  });
});
