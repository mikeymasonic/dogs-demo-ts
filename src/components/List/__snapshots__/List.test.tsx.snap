// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component it should mock the context and render List 1`] = `ShallowWrapper {}`;
